# Case Study - Customer Retention

<center><img src="https://1000logos.net/wp-content/uploads/2023/04/Starbucks-logo.png" height=216, width=384></img></center>

#### You are a dedicated data analyst at Starbucks in Kuala Lumpur, Malaysia, have noticed a recent fluctuation in customer loyalty among in-store patrons. Your goal is to not only enhance customer satisfaction as a factor of customer retention but also pave the way for increased revenue.

## Business Understanding

### Defining the Problem Statements

Using SMART framework:

1. **Specific**: Enhanching loyalty among existing customers especially in store.

2. **Measurable**: Achieve 90% customer retention rate.

3. **Achievable**: Increasing quality of product, service, in-store ambience along with the WIFI quality. Also, implement promos.

4. **Relevant**: Increasing the customer retention can lead to achieve higher revenue.

5. **Time-Bound**: Achieve within the next quarter.

`Problem statement`:


The goal is to enhance customer loyalty among existing in-store customers, aiming for a 90% customer retention rate within the next quarter. This will be achieved by improving product and service quality, enhancing in-store ambience, optimizing WIFI quality, and implementing strategic promotions, ultimately driving higher revenue.


### Breaking Down the Problem

Main problem: `improving customer retention as a metric of customer loyalty`

To ease our analysis and solve the problem, we need to understand the detail about the problem. To do it, we can use any framework to find the problem's root such as `5W+1H`, `Fish Bond Diagram`, etc. However, we will use `5W+1H`.

The `5W+1H`s:
- What factors that can lead to customer retention improvement?
- Who are people that be loyal to Starbucks?
- How will the quality of products and services be enhanced to improve customer retention?
- Why is improving customer retention important for company?
- etc

Moreover, we only focus to answer the top 3 questions to guide our analysis

## Data Understanding

### Basic Data Information

**Dataset Description**

> This dataset is composed of a survey questions of over 100 respondents for their buying behavior at Starbucks.
Income is show in Malaysian Ringgit (RM)

**Context**

> Predict behavior to retain customers. You can analyze all relevant customer data and develop focused customer retention programs

**Content**

> Demographic info about customers – gender, age range, employment status, income range
Their current behavior in buying Starbucks
Facilities and features of Starbucks that contribute to the behavior

**Columns**
```
Timestamp
1. Your Gender
2. Your Age
3. Are you currently....?
4. What is your annual income?
5. How often do you visit Starbucks?
6. How do you usually enjoy Starbucks?
7. How much time do you normally  spend during your visit?
8. The nearest Starbucks's outlet to you is...?
9. Do you have Starbucks membership card?
10. What do you most frequently purchase at Starbucks?
11. On average, how much would you spend at Starbucks per visit?
12. How would you rate the quality of Starbucks compared to other brands (Coffee Bean, Old Town White Coffee..) to be:
13. How would you rate the price range at Starbucks?
14. How important are sales and promotions in your purchase decision?
15. How would you rate the ambiance at Starbucks? (lighting, music, etc...)
16. You rate the WiFi quality at Starbucks as..
17. How would you rate the service at Starbucks? (Promptness, friendliness, etc..)
18. How likely you will choose Starbucks for doing business meetings or hangout with friends?
19. How do you come to hear of promotions at Starbucks? Check all that apply.
20. Will you continue buying at Starbucks?
```

In [ ]:
import pandas as pd
from scipy import stats
import requests

Our data is from an API that you can access with this url: https:/p0w3casestudy.vercel.app/

As we learned on day 2, to fetch data from API we can use requests along with `get` endpoint. In addition, we need to check the data type of the result. Is it in json, string, or something else.

However, our result is dictionary, so we can directly convert into Pandas Data Frame.

In [ ]:
url = "https://p0w3casestudy.vercel.app/"
result = requests.get(url)
print(type(result.json()))

<class 'dict'>


In [ ]:
data = pd.DataFrame(result.json())
data.head()

Timestamp  Gender            Age Occupation  \
0  2019/10/01 12:38:43 PM GMT+8  Female  From 20 to 29    Student   
1  2019/10/01 12:38:54 PM GMT+8  Female  From 20 to 29    Student   
2  2019/10/01 12:38:56 PM GMT+8    Male  From 20 to 29   Employed   
3  2019/10/01 12:39:08 PM GMT+8  Female  From 20 to 29    Student   
4  2019/10/01 12:39:20 PM GMT+8    Male  From 20 to 29    Student   

        Annual Income    Visit  Behaviour                    Time Spend  \
0  Less than RM25,000   Rarely    Dine in  Between 30 minutes to 1 hour   
1  Less than RM25,000   Rarely  Take away              Below 30 minutes   
2  Less than RM25,000  Monthly    Dine in  Between 30 minutes to 1 hour   
3  Less than RM25,000   Rarely  Take away              Below 30 minutes   
4  Less than RM25,000  Monthly  Take away  Between 30 minutes to 1 hour   

        Distance Membership  ...         Money Spend Quality Rate  Price Rate  \
0     within 1km        Yes  ...      Less than RM20            4           3   
1      1km - 3km        Yes  ...      Less than RM20            4           3   
2  more than 3km        Yes  ...      Less than RM20            4           3   
3  more than 3km         No  ...      Less than RM20            2           1   
4      1km - 3km         No  ...  Around RM20 - RM40            3           3   

   Importance of Promo  Ambience Rate  WIFI Rate  Service Rate  \
0                    5              5          4             4   
1                    4              4          4             5   
2                    4              4          4             4   
3                    4              3          3             3   
4                    4              2          2             3   

   Preference Rate                                  Promotion Channel  \
0                3  Starbucks Website/Apps;Social Media;Emails;Dea...   
1                2                     Social Media;In Store displays   
2                3                       In Store displays;Billboards   
3                3                  Through friends and word of mouth   
4                3                Starbucks Website/Apps;Social Media   

  Buying Again  
0          Yes  
1          Yes  
2          Yes  
3           No  
4          Yes  

[5 rows x 21 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 122 entries, 0 to 121
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Timestamp            122 non-null    object
 1   Gender               122 non-null    object
 2   Age                  122 non-null    object
 3   Occupation           122 non-null    object
 4   Annual Income        122 non-null    object
 5   Visit                122 non-null    object
 6   Behaviour            121 non-null    object
 7   Time Spend           122 non-null    object
 8   Distance             122 non-null    object
 9   Membership           122 non-null    object
 10  Item Purchase        122 non-null    object
 11  Money Spend          122 non-null    object
 12  Quality Rate         122 non-null    int64 
 13  Price Rate           122 non-null    int64 
 14  Importance of Promo  122 non-null    int64 
 15  Ambience Rate        122 non-null    int64 
 16  WIFI Rate    

Our data consists of 21 columns and 122 rows. Pretty small for data analysis but it's okey for learning and initial approachment. We have inferential statistics, lol.

Unfortunately, we have missing values in our data which are in `Behaviour` and `Promotion Channel` columns. We will take out the row(s) later on.

## Data Preparation

Foretunately, our data is almost perfect to perform data analysis, so we only need to handle the missing values. According to our discussion above, we only need to take out the missing values.

However, how about the `Timestamp` column? do we need to convert the data type into `datetime64[ns]`? Hmm, it's not important to do that since the `Timestamp` relates to the record time in the databas. Remember that, our data actually is not time series data, each row represents one customer. Alternatively, we can takout the `Timestamp` from our data.

### Missing Values Handling

In [ ]:
data = data.dropna()
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 121 entries, 0 to 121
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Timestamp            121 non-null    object
 1   Gender               121 non-null    object
 2   Age                  121 non-null    object
 3   Occupation           121 non-null    object
 4   Annual Income        121 non-null    object
 5   Visit                121 non-null    object
 6   Behaviour            121 non-null    object
 7   Time Spend           121 non-null    object
 8   Distance             121 non-null    object
 9   Membership           121 non-null    object
 10  Item Purchase        121 non-null    object
 11  Money Spend          121 non-null    object
 12  Quality Rate         121 non-null    int64 
 13  Price Rate           121 non-null    int64 
 14  Importance of Promo  121 non-null    int64 
 15  Ambience Rate        121 non-null    int64 
 16  WIFI Rate    

Clean from missing values. We only lost 1 row.

In [ ]:
df = data.drop(columns='Timestamp')
df.head()

Gender            Age Occupation       Annual Income    Visit  Behaviour  \
0  Female  From 20 to 29    Student  Less than RM25,000   Rarely    Dine in   
1  Female  From 20 to 29    Student  Less than RM25,000   Rarely  Take away   
2    Male  From 20 to 29   Employed  Less than RM25,000  Monthly    Dine in   
3  Female  From 20 to 29    Student  Less than RM25,000   Rarely  Take away   
4    Male  From 20 to 29    Student  Less than RM25,000  Monthly  Take away   

                     Time Spend       Distance Membership  \
0  Between 30 minutes to 1 hour     within 1km        Yes   
1              Below 30 minutes      1km - 3km        Yes   
2  Between 30 minutes to 1 hour  more than 3km        Yes   
3              Below 30 minutes  more than 3km         No   
4  Between 30 minutes to 1 hour      1km - 3km         No   

          Item Purchase         Money Spend  Quality Rate  Price Rate  \
0                Coffee      Less than RM20             4           3   
1  Cold drinks;Pastries      Less than RM20             4           3   
2                Coffee      Less than RM20             4           3   
3                Coffee      Less than RM20             2           1   
4     Coffee;Sandwiches  Around RM20 - RM40             3           3   

   Importance of Promo  Ambience Rate  WIFI Rate  Service Rate  \
0                    5              5          4             4   
1                    4              4          4             5   
2                    4              4          4             4   
3                    4              3          3             3   
4                    4              2          2             3   

   Preference Rate                                  Promotion Channel  \
0                3  Starbucks Website/Apps;Social Media;Emails;Dea...   
1                2                     Social Media;In Store displays   
2                3                       In Store displays;Billboards   
3                3                  Through friends and word of mouth   
4                3                Starbucks Website/Apps;Social Media   

  Buying Again  
0          Yes  
1          Yes  
2          Yes  
3           No  
4          Yes

Created new dataframe without `Timestamp`

## Modeling/Exploratory Data Analysis

#### What factors that can lead to customer retention improvement?

To answer this question, we will perform chi-squared test since we want to test the relation between two categorical data. But, how about rating data? they are "numerical" data aren't they? Moreover, we will treat the rating data as categorical data to ease our works.

To perform the test, we need to define the hypothesis:

$H_0$: A variable and `buying again` are independent

$H_1$: A variable and `buying again` are dependent

We will use confidence level of 95%.

In [ ]:
cols = data.drop(columns=['Timestamp','Promotion Channel','Buying Again']).columns

for col in cols:
  cross = pd.crosstab(data[col],data['Buying Again'])
  pval = stats.chi2_contingency(cross).pvalue
  prompt = "and they are dependent" if pval<0.05 else ""
  print(f"P-value of {col} and Buying Again: {pval} {prompt}")

P-value of Gender and Buying Again: 1.0 
P-value of Age and Buying Again: 0.6356871673042823 
P-value of Occupation and Buying Again: 0.1607051559293076 
P-value of Annual Income and Buying Again: 0.7438495801427483 
P-value of Visit and Buying Again: 0.0081671269342487 and they are dependent
P-value of Behaviour and Buying Again: 0.1968218965282764 
P-value of Time Spend and Buying Again: 0.07096869017305363 
P-value of Distance and Buying Again: 0.5201380402399829 
P-value of Membership and Buying Again: 0.0005711760401142383 and they are dependent
P-value of Item Purchase and Buying Again: 0.5439062916107664 
P-value of Money Spend and Buying Again: 0.00011367023704496573 and they are dependent
P-value of Quality Rate and Buying Again: 6.93854822279136e-05 and they are dependent
P-value of Price Rate and Buying Again: 1.45956480340212e-05 and they are dependent
P-value of Importance of Promo and Buying Again: 0.853345487868004 
P-value of Ambience Rate and Buying Again: 0.0061673913

Based on this, the factors that relate to customer retention are visit intensity, membership, money spend, product quality, price, ambience, service quality, and purpose of visit/preference.

However, the detail of which one leads the customers to be loyal, you need to explore each variable.

Let's we encode column `Buying Again` to make retention rate calculation easier.

In [ ]:
df = data.copy()
df['Buying Again'] = df['Buying Again'].replace({'Yes':1,'No':0})
df['Buying Again'].head()

0    1
1    1
2    1
3    0
4    1
Name: Buying Again, dtype: int64

To explore how the variable and retention relation works, we need to calculate the retention rate at first by measure the average of 'Buying Again'. The mean is the retention rate!

In [ ]:
df.groupby('Visit')[['Buying Again']].mean()

Buying Again
Visit                
Daily        1.000000
Monthly      0.961538
Never        0.500000
Rarely       0.710526
Weekly       1.000000

We see that actually visit intensity to the store is not really affect the customer retention.

In [ ]:
df.groupby('Membership')[['Buying Again']].mean()

Buying Again
Membership              
No              0.639344
Yes             0.916667

Also, if customers has membership or not, they will keep to be loyal to Starbucks

In [ ]:
df.groupby('Money Spend')[['Buying Again']].mean()

Buying Again
Money Spend                     
Around RM20 - RM40      0.933333
Less than RM20          0.706897
More than RM40          1.000000
Zero                    0.363636

People that spend their money to buy starcuks' products tend to be loyal and will buying again.

Rating data are actually ordinal data, so to measure the detail relation, we can use kendall correlation.

In [ ]:
df.groupby('Quality Rate')[['Buying Again']].mean().reset_index()

Quality Rate  Buying Again
0             1      0.000000
1             2      0.250000
2             3      0.743590
3             4      0.857143
4             5      0.913043

In [ ]:
def corr(col):
  tmp = df.groupby(col)[['Buying Again']].mean().reset_index()
  tau, pval = stats.kendalltau(tmp[col],tmp['Buying Again'])
  print('Kendall Tau:',tau)
  print('P-value:',pval)

In [ ]:
corr('Quality Rate')

Kendall Tau: 0.9999999999999999
P-value: 0.016666666666666666


In [ ]:
corr('Price Rate')

Kendall Tau: 0.9999999999999999
P-value: 0.016666666666666666


In [ ]:
corr('Ambience Rate')

Kendall Tau: 0.9486832980505137
P-value: 0.022977401503206065


In [ ]:
corr('Service Rate')

Kendall Tau: 0.7999999999999999
P-value: 0.08333333333333333


In [ ]:
corr('Preference Rate')

Kendall Tau: 0.9999999999999999
P-value: 0.016666666666666666


Based on those corr values, we know that retention rate has high correlation to customer satisfactory against product quality, price, ambience, service quality.

And because they are satisfy to Starbucks, they will come to buy the products again and willing to hangout with their friends or doing business meeting in Starbucks.

However, only Service rate that shows of p-value more than critical value, 0.05, so mostly, the correlations are occured not by chance randomly and we need more data to prove it. Moreover, we can still continue our analysis and this can be a warning that our conclusion maybe wrong.

#### Who are people that be loyal to Starbucks?

In [ ]:
loyals = data[data['Buying Again']=='Yes']
loyals.head()

Timestamp  Gender            Age Occupation  \
0  2019/10/01 12:38:43 PM GMT+8  Female  From 20 to 29    Student   
1  2019/10/01 12:38:54 PM GMT+8  Female  From 20 to 29    Student   
2  2019/10/01 12:38:56 PM GMT+8    Male  From 20 to 29   Employed   
4  2019/10/01 12:39:20 PM GMT+8    Male  From 20 to 29    Student   
5  2019/10/01 12:39:39 PM GMT+8  Female  From 20 to 29    Student   

        Annual Income    Visit  Behaviour                    Time Spend  \
0  Less than RM25,000   Rarely    Dine in  Between 30 minutes to 1 hour   
1  Less than RM25,000   Rarely  Take away              Below 30 minutes   
2  Less than RM25,000  Monthly    Dine in  Between 30 minutes to 1 hour   
4  Less than RM25,000  Monthly  Take away  Between 30 minutes to 1 hour   
5  Less than RM25,000   Rarely    Dine in  Between 30 minutes to 1 hour   

        Distance Membership  ...         Money Spend Quality Rate  Price Rate  \
0     within 1km        Yes  ...      Less than RM20            4           3   
1      1km - 3km        Yes  ...      Less than RM20            4           3   
2  more than 3km        Yes  ...      Less than RM20            4           3   
4      1km - 3km         No  ...  Around RM20 - RM40            3           3   
5  more than 3km         No  ...      Less than RM20            4           3   

   Importance of Promo  Ambience Rate  WIFI Rate  Service Rate  \
0                    5              5          4             4   
1                    4              4          4             5   
2                    4              4          4             4   
4                    4              2          2             3   
5                    5              5          4             5   

   Preference Rate                                  Promotion Channel  \
0                3  Starbucks Website/Apps;Social Media;Emails;Dea...   
1                2                     Social Media;In Store displays   
2                3                       In Store displays;Billboards   
4                3                Starbucks Website/Apps;Social Media   
5                4                                       Social Media   

  Buying Again  
0          Yes  
1          Yes  
2          Yes  
4          Yes  
5          Yes  

[5 rows x 21 columns]

In [ ]:
len(loyals)

94

There are 94 out of 121 people that willing to buy again Starbucks products

In [ ]:
df['Gender'].value_counts()

Female    65
Male      56
Name: Gender, dtype: int64

Mostly female

In [ ]:
df['Age'].value_counts()

From 20 to 29    84
From 30 to 39    17
Below 20         13
40 and above      7
Name: Age, dtype: int64

They are youngsters

In [ ]:
df['Occupation'].value_counts()

Employed         60
Student          42
Self-employed    17
Housewife         2
Name: Occupation, dtype: int64

Mostly they have already employed

In [ ]:
df['Annual Income'].value_counts()

Less than RM25,000       70
RM25,000 - RM50,000      25
RM50,000 - RM100,000     17
More than RM150,000       6
RM100,000 - RM150,000     3
Name: Annual Income, dtype: int64

Mostly their salary not that big (less then RM 25000)

In [ ]:
df['Visit'].value_counts()

Rarely     76
Monthly    26
Weekly      9
Never       8
Daily       2
Name: Visit, dtype: int64

They are not often to come to Starbucks

In [ ]:
df['Behaviour'].value_counts()

Take away             49
Dine in               46
Drive-thru            20
never                  2
Never buy              1
I dont like coffee     1
Never                  1
Never                  1
Name: Behaviour, dtype: int64

Mostly take away and dine-in

In [ ]:
df['Time Spend'].value_counts()

Below 30 minutes                72
Between 30 minutes to 1 hour    34
Between 1 hour to 2 hours       12
More than 3 hours                2
Between 2 hours to 3 hours       1
Name: Time Spend, dtype: int64

They don't spend much time in store (perhaps becuase of most of them take away)

In [ ]:
df['Membership'].value_counts()

No     61
Yes    60
Name: Membership, dtype: int64

Possessing the membership or not, it hasn't any difference

In [ ]:
df['Money Spend'].value_counts()

Less than RM20        58
Around RM20 - RM40    45
Zero                  11
More than RM40         7
Name: Money Spend, dtype: int64

Most of them willing to spend less than RM 20. Buying the cheapest ones

From this analysis, we can deduce that most of the loyal customers are young individuals with relatively low incomes, tending to be impulsive buyers and more inclined to purchase inexpensive products.

#### How will the quality of products and services be enhanced to improve customer retention?

In [ ]:
#Products Quality
df.groupby('Quality Rate')[['Buying Again']].mean().reset_index()

Quality Rate  Buying Again
0             1      0.000000
1             2      0.250000
2             3      0.743590
3             4      0.857143
4             5      0.913043

In [ ]:
corr("Quality Rate")

Kendall Tau: 0.9999999999999999
P-value: 0.016666666666666666


In [ ]:
#Products Quality
df.groupby('Service Rate')[['Buying Again']].mean().reset_index()

Service Rate  Buying Again
0             1      0.000000
1             2      0.500000
2             3      0.785714
3             4      0.725490
4             5      0.956522

In [ ]:
corr("Service Rate")

Kendall Tau: 0.7999999999999999
P-value: 0.08333333333333333


We've already know that higher product quality and service rate, can increasing the customer retention rate. This because more people satisfy with the products and service, more willing to come again and buy the products again.

## Conclusion

In conclusion, improving customer retention can be done by enhancing customer satisfactory. In aim to enhanching customer satisfactory, we can improve quality of service and products, improve the in-store ambience since our loyal customers also mostly come to dine-in. Also, Most our loyal customers are youngsters that has small salary, tend to be more consumptive, and willing to pay less.